kolibri/content/api.py

```diff
@@ -103,26 +103,55 @@ def filter_search(self, queryset, name, value):
         Implement various filtering strategies in order to get a wide range of search results.
         """
 
-        # search the whole string in the title or description
-        query = Q(title__icontains=value) | Q(description__icontains=value)
-
-        # split string up by spaces and search each individual string in title or description
-        str_arr = value.split()
-        for string in str_arr:
-            query.add(Q(title__icontains=string) | Q(description__icontains=string), Q.OR)
-        matches = queryset.filter(query)
-
-        # remove duplicate content items based on content_id
-        content_id_set = set()
-        id_list = []
-        for id_pair in matches.values('content_id', 'id'):
-            if id_pair['content_id'] in content_id_set:
-                pass
-            else:
-                content_id_set.add(id_pair['content_id'])
-                id_list.append(id_pair['id'])
-
-        return matches.filter(id__in=id_list[:30])
+        # return the result of and-ing a list of queries
+        def intersection(queries):
+            return reduce(lambda x, y: x & y, queries)
+
+        # return the result of or-ing a list of queries
+        def union(queries):
+            return reduce(lambda x, y: x | y, queries)
+
+        # queries ordered by relevance priority
+        words = value.split()
+        all_queries = [
+            # exact match in title
+            Q(title__icontains=value),
+            # all words in title
+            intersection([Q(title__icontains=w) for w in words]),
+            # exact match in description
+            Q(description__icontains=value),
+            # all words in description
+            intersection([Q(description__icontains=w) for w in words]),
+        ]
+
+        # any word in title, sorted by length
+        for w in reversed(sorted(words, key=len)):
+            all_queries.append(Q(title__icontains=w))
+
+        results = []
+        content_ids = set()
+        MAX_RESULTS = 30
+        BUFFER_SIZE = MAX_RESULTS * 3  # grab some extras, but not too many
+
+        # iterate over each query type, and build up search results
+        for query in all_queries:
+            # in each pass, don't take any items already in the result set
+            matches = queryset.exclude(content_id__in=list(content_ids)).filter(query)[:BUFFER_SIZE]
+
+            for match in matches:
+                # filter the dupes
+                if match.content_id in content_ids:
+                    continue
+                # add new, unique results
+                content_ids.add(match.content_id)
+                results.append(match)
+
+                # bail out as soon as we reach the quota
+                if len(results) >= MAX_RESULTS:
+                    return results
```

Comment:
Adding the fuzzed results back in as the last "query" could help fill in the gaps if there are typos.

Reply:
Or maybe it should be handled specially -- and only be called when there are *no* exact results. And pass a flag back indicating that it's a fuzzy set of results, so the UI can indicate "No exact results found for your search. Some possible matches are shown below."

Reply:
sounds reasonable .. but I think the stemmed metaphone stuff was english-only?

Reply:
created a new issue here: https://github.com/learningequality/kolibri/issues/3684